### PR TITLE
MediaRanker - Hannah

### PR DESCRIPTION
**Pull request by hertweckhr1**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote self.top_ten custom method in order to sort the data by number of votes in descending order and grab the top ten from each specific category.
Describe how you approached testing that model method. What edge cases did you come up with? | Since I used the same method style for albums, movies, and books, I tested nominally, that it would accumulate an array no greater than ten and that the first value in the array would be greater than or equal to the last value in the array.  I also tested as an edge, when there were zero works in that category and if there was only one work in that category. 
What are `session` and `flash`? What is the difference between them? | Session is used to save information over a users session keeping track of activity once logged in through cookies. Flash is a one time message shown to the users telling the user something happened.  It is also a way to display error messages. 
Describe a controller filter you wrote. | I wrote a controller filter for find_work in order to dry up my works controller. I included it in the private section of my works controller and included before_action :find_work, only: [:show, :edit, :update, :destroy, :upvote] at the very top of the controller to make sure this action ran before the following methods. 
What was one thing that you gained more clarity on through this assignment? | I continued to gain more clarity on how all the pieces worked together.  Additionally, gained more clarity on flash messaging, writing tests, and sessions. All of which, I hope to gain much more clarity with further projects.
What is the Heroku URL of your deployed application? | https://hannah-mediaranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |  Since we learned more on bootstrap this week, it would have been fun to start experimenting with it for this project in a wireframe set up rather than copying an existing website.

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
```
Comment (CheezItMan) [inline app/models/vote.rb:4]: You should have a validation to keep a vote from being created with a duplicate combination of work and user.

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  validates :title, uniqueness: true, presence: true
+  validates :creator, presence: true
+  validates :publication_year, presence: true,
+    numericality: { only_integer: true, greater_than: 0 },
+    length: { is: 4 }
+  validates :description, presence: true
+  validates :category, presence: true
+
+
+  def self.sorted_works(category)
```
Comment (CheezItMan) [inline app/models/work.rb:14]: 💯

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    works = Work.where(category: category)
+    sorted_works = works.sort_by { |work| work.votes.count }
```
Comment (CheezItMan) [inline app/models/work.rb:16]: To keep form having to reverse the sort, you can:

```ruby
 works.sort_by { |work| - work.votes.count }
```

File: config/routes.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
+  post '/login', to: 'sessions#login', as: 'login'
+  get '/login', to: 'sessions#new', as: 'new_session'
+  delete 'sessions/destroy', to: 'sessions#destroy', as: 'logout'
+
+  resources :users, only: [:index, :show]
+  resources :works
+  post 'works/:id/upvote', to: 'works#upvote', as: 'upvote'
+
```
Comment (CheezItMan) [inline config/routes.rb:9]: 💯 🥇

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      valid = user.valid?
+
+      expect(valid).must_equal false
+      expect(user.errors.messages).must_include :user_name
+    end
+
+    it 'must have a title with a min of 5 letters' do
+      #arrange
+      user.user_name = ''
+      4.times do
```
Comment (CheezItMan) [inline test/models/user_test.rb:46]: After this loop add one more letter and verify that the user_name then works.

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+        valid = user.valid?
+
+        expect(valid).must_equal false
+        expect(user.errors.messages).must_include :user_name
+      end
+
+    end
+
+    it 'must be no longer than 20 characters for username' do
```
Comment (CheezItMan) [inline test/models/user_test.rb:57]: You should also verify that a username with 19 letters is valid.

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    it 'must have a numerical year published' do
+      work.publication_year = 'seaweed'
+
+      valid = work.valid?
+
+      expect(valid).must_equal false
+      expect(work.errors.messages).must_include :publication_year
+    end
+
+    it "must have a year with four digits" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:79]: You should test that the `publication_year` is invalid with 3 digits, valid with 4 digits and invalid with 5.

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+      work.publication_year = 87
+
+      valid = work.valid?
+
+      expect(valid).must_equal false
+      expect(work.errors.messages).must_include :publication_year
+    end
+
+    it 'must have a category' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:88]: You should also limit the potential categories.

File: app/views/works/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            <td><%= album.publication_year %></td>
+            <td><%= link_to "Upvote", upvote_path(album.id), class: "btn btn-primary", rel: "nofollow", method: "post" %></td>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+  </section>
+  <section class="media-table">
+    <h4>Books</h4>
+    <table class="table">
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:33]: Notice how these tables are repetitive.  These could be DRYed up with a partial.

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages.
Comprehension questions	|  Check, I'm glad you gained more clarity.  
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Well done
Views are well-organized (DRY, use of semantic HTML, use of partials) |  You could make better use of partials to DRY up your views.  
Errors are reported to the user |  Check
Business logic lives in the models |  Check
Models are thoroughly tested, including relations, validations and any custom logic |  You have some pretty good tests, I left a few comments however.  
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |  Check
**Wave 2 - Users and Votes** |
Users can log in and log out |  Check
The ID of the current user is stored in the session | Check
A user cannot vote for the same media more than once | Check
All media lists are ordered by vote count | Check
Splash page contains a media spotlight | Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | Check
Individual user pages and the user list are present | Check
**Optional - Styling** |
Bootstrap is used appropriately |  Very nice styling
Look and feel is similar to the original |  It looks very very close
**Overall** |  Nice work, you hit all the learning goals.  One suggestion is to make better use of partials to DRY up views.  Check out my comments, and let me know if you have any questions.
